### ISSUE TITLE
Lifetime cannot be elided

### ISSUE THREAD

**Issue by samueltardieu** [rust-lang/rust-clippy#7893]
Lint name: clippy::needless_lifetimes

I tried this code:

```rust
async fn func<'a>(args: &[&'a str]) -> Option<&'a str> {
    args.get(0).cloned()
}
```

I expected to see this happen: no warning as the `'a` lifetime is needed. Moreover the lint does not trigger is the function is not marked as `async`.

Instead, this happened: 

```
warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
 --> src/main.rs:1:1
  |
1 | async fn func<'a>(args: &[&'a str]) -> Option<&'a str> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::needless_lifetimes)]` on by default
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (4e0d3973f 2021-10-27)
binary: rustc
commit-hash: 4e0d3973fafdfb1c51011bc74e44257b5e3863f1
commit-date: 2021-10-27
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```
